build-system/tasks/dep-check.js

```diff
@@ -0,0 +1,159 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+var BBPromise = require('bluebird');
+var babel = require('babelify');
+var browserify = require('browserify');
+var buffer = require('vinyl-buffer');
+var fs = BBPromise.promisifyAll(require('fs-extra'));
+var gulp = require('gulp-help')(require('gulp'));
+var minimist = require('minimist');
+var path = require('path');
+var source = require('vinyl-source-stream');
+var through = require('through2');
+var util = require('gulp-util');
+
+
+var argv = minimist(process.argv.slice(2), { boolean: ['strictBabelTransform'] });
+
+
+/**
+ * @typedef {{
+ *   name: string,
+ *   deps: !Array<!Object<string, string>
+ * }}
+ */
+var DirtyDepDef;
+
+/**
+ * Returns a list of extension folders
+ * @return {!Promise<!Array<string>>}
+ */
+function getExtensions() {
+  var root = process.cwd();
+  return fs.readdirAsync('extensions').then((dirItems) => {
+    return dirItems
+      .map((item) => `${root}/extensions/${item}`)
+      .filter((item) => fs.statSync(item).isDirectory())
+  });
+}
+
+/**
+ * @param {string} entryModule
+ * @return {!Promise<!DirtyDepDef>}
+ */
+function getGraph(entryModule) {
+  var resolve;
+  var promise = new BBPromise((r) => {
+    resolve = r;
+  });
+  var module = Object.create(null);
+  module.name = entryModule.replace(`${root}/`, '');
+  module.deps = [];
+
+  var hasher = through.obj(function (row, enc, next) {
+      module.deps.push(row.deps);
+      this.push(row);
+      next();
+  });
+
+  var bundler = browserify(entryModule, {debug: true, deps: true})
+      .transform(babel, {
+        loose: argv.strictBabelTransform ? undefined : 'all'
+      });
+
+  bundler.pipeline.get('deps').push(hasher);
+  bundler.bundle()
+      .pipe(source(entryModule))
+      .pipe(gulp.dest('./.amp-build'))
+      .on('end', resolve.bind(null, module));
+  return promise;
+}
+
+/**
+ * @return {string}
+ */
+function getEntryModule(extensionFolder) {
+  // TODO: handle more than just 0.1
+  return `${extensionFolder}/0.1/${path.basename(extensionFolder)}.js`;
+}
+
+/**
+ * @param {!Array<!DirtyDepDef>} deps
+ * @return {!Object<string, !Array>}
+ */
+function getCleanGraph(deps) {
+    var root = process.cwd();
+    var modules = Object.create(null);
+    deps.forEach((x) => {
+      // Flatten deps
+      // Remove empty objects
+      // Remove the relative path import keys
+      var deps = [].concat.apply([], x.deps)
+          .filter((x) => Object.keys(x).length)
+          .map((x) => Object.keys(x).map((y) => x[y]));
+      // Make list unique
+      // Make project folder as root
+      deps = deps.concat.apply([], deps)
+          .filter((x, i, a) => a.indexOf(x) === i)
+          .map((x) => x.replace(`${root}/`, ''));
+      modules[x.name] = deps;
+    })
+    return modules;
+}
+
+/**
+ * @param {!Object<string, !Array>} modules
+ */
+function checkImplImport(modules) {
```

Comment:
This PR does the heavy lifting and we can submit it like this, but what I'm hoping for is to have a declarative way to configure forbidden deps.

E.g.

```js
const rules = [
  {
    type: 'forbidden',
    filesMatching: 'extensions/**/*.js',
    mustNotDependOn: 'src/services/**/*.js',
    whitelist: [
      ['foo/bar.js', 'src/service/cid.js']
    ]
  }
]
```